### PR TITLE
Improve hash function based on whole template

### PR DESCRIPTION
**Pull request by AverageMarcus**
Signed-off-by: Marcus Noble <github@marcusnoble.co.uk>

### What this PR does / why we need it

Updates the usage of the `hash` function to pass in the whole template its related to, rather than just the values. 

This should hopefully avoid all instances where Helm attempts to update immutable values in-place rather than recreating.

### Checklist

- [x] Update changelog in CHANGELOG.md.

### Trigger e2e tests

/test create

### CODE REVIEW

File: helm/cluster-gcp/templates/_bastion.tpl [hunk 1]
```diff
@@ -43,34 +72,21 @@ metadata:
   labels:
     cluster.x-k8s.io/role: bastion
     {{- include "labels.common" $ | nindent 4 }}
-  name: {{ include "resource.default.name" $ }}-bastion-{{ include "hash" (dict "data" .Values.bastion "global" .) }}
+  name: {{ include "resource.default.name" $ }}-bastion-{{ include "hash" (dict "data" (include "bastion-machinetemplate-spec" $) "global" .) }}
```
Comment (calvix) [inline helm/cluster-gcp/templates/_bastion.tpl:75]: but does this actually change when values change as well? is the hash computed after all values are filled?

**Comment by calvix** [comment]
I think we will need this for AWS as well, should I create a ticket ?

File: helm/cluster-gcp/templates/_bastion.tpl (hunk 1 above)
Comment (AverageMarcus) [reply helm/cluster-gcp/templates/_bastion.tpl]: Yes. It's based on the rendered template

Comment (calvix) [reply helm/cluster-gcp/templates/_bastion.tpl]: ack

**Review by calvix — APPROVED** [review]
happy if you are happy

**Review by mnitchev — APPROVED** [review]

File: helm/cluster-gcp/templates/_machine_deployments.tpl [hunk 2]
```diff
@@ -45,62 +106,23 @@ metadata:
...
-      files:
-      {{- include "sshFiles" $ | nindent 6 }}
-      {{- include "diskFiles" $ | nindent 6 }}
-      preKubeadmCommands:
-      - /bin/bash /opt/init-disks.sh
-      postKubeadmCommands:
-      {{- include "sshPostKubeadmCommands" . | nindent 6 }}
-      users:
-      {{- include "sshUsers" . | nindent 6 }}
+    spec:{{ include "machinedeployment-kubeadmconfigtemplate-spec" $thisVal | nindent 6 }}
```
Comment (AverageMarcus) [inline helm/cluster-gcp/templates/_machine_deployments.tpl:125]:
```suggestion
    spec: {{ include "machinedeployment-kubeadmconfigtemplate-spec" $thisVal | nindent 6 }}
```

**Pull request merged**